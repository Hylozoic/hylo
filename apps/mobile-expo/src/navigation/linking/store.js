import { create } from 'zustand'

const useLinkingStore = create((set) => ({
  initialURL: null,
  setInitialURL: (url) => set({ initialURL: url }),
  returnToOnAuthPath: null,
  setReturnToOnAuthPath: (path) => set({ returnToOnAuthPath: path })
}))

export default useLinkingStore