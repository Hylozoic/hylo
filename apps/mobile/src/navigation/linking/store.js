import { create } from 'zustand'

const useLinkingStore = create((set) => ({
  initialURL: null,
  returnToOnAuthPath: null,
  setInitialURL: initialURL => set({ initialURL }),
  setReturnToOnAuthPath: returnToOnAuthPath => set({ returnToOnAuthPath })
}))

export default useLinkingStore
