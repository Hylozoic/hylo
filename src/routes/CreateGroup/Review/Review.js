import React, { Component, createRef } from 'react'
import '../CreateGroup.scss'
import ModalSidebar from 'components/ModalSidebar'
import TextInput from 'components/TextInput'
import { hyloNameWhiteBackground, groovingAxolotl } from 'util/assets'
import { bgImageStyle } from 'util/index'
import ModalFooter from 'components/ModalFooter'
import { find } from 'lodash'
import { get } from 'lodash/fp'
import {
  slugValidatorRegex,
  formatDomainWithUrl,
  removeUrlFromDomain,
  invalidSlugMessage
} from '../util'

export default class Review extends Component {
  constructor () {
    super()
    const fields = ['groupName', 'groupDomain']
    this.inputRefs = fields.reduce((acc, name) => {
      acc[name] = createRef()
      return acc
    }, {})
    this.state = {
      edits: fields.reduce((acc, name) => {
        acc[name] = null
        return acc
      }, {}),
      readOnly: fields.reduce((acc, name) => {
        acc[name] = true
        return acc
      }, {})
    }
  }

  editHandler = (name) => {
    this.setState({
      readOnly: {
        ...this.state.readOnly,
        [name]: false
      }
    })
    this.inputRefs[name].current.select()
  }

  handleInputChange = (event, name) => {
    let value = event.target.value
    if (name === 'groupDomain') {
      value = removeUrlFromDomain(value)
      if (value !== '') {
        this.props.fetchGroupExists(value)
      }
    }
    this.setState({
      edits: {
        ...this.state.edits,
        [name]: value,
        changed: true
      }
    })
  }

  onEnter = event => {
    if (event.key === 'Enter') {
      this.errorCheckAndSubmit()
    }
  }

  submit = () => {
    const { groupName, groupDomain, groupNetworkId } = this.state.edits
    this.props.createGroup(
      // groupPrivacy,
      groupName,
      groupDomain,
      groupNetworkId
    )
      .then(({ error }) => {
        if (error) {
          this.setState({
            error: 'There was an error, please try again.'
          })
        } else {
          this.props.goToGroup(`${groupDomain}`)
        }
      })
  }

  errorCheckAndSubmit = () => {
    const { groupName, groupDomain } = this.state.edits
    if (groupName === '' || groupDomain === '') {
      this.setState({
        error: 'Please fill in each field.'
      })
    } else if (this.props.groupDomainExists) {
      this.setState({
        error: 'This URL already exists. Try another.'
      })
    } else if (!slugValidatorRegex.test(removeUrlFromDomain(groupDomain))) {
      formatDomainWithUrl(groupDomain)
      this.setState({
        error: invalidSlugMessage
      })
    } else {
      this.submit()
    }
  }

  componentWillMount = () => {
    const { groupPrivacy } = this.props
    const privacyOption = find(privacyOptions, { label: groupPrivacy })
    const selectedGroupPrivacy = get('label', privacyOption) // set to Private by default
    this.setState({
      edits: {
        groupName: get('groupName', this.props) || '',
        groupDomain: get('groupDomain', this.props) || '',
        groupNetworkId: get('groupNetworkId', this.props) || '',
        groupPrivacy: selectedGroupPrivacy
      }
    })
  }

  formatDomainWithUrl (groupDomain) {
    if (!groupDomain) return null
    let formattedDomain = groupDomain.replace('hylo.com/g/', '').replace('hylo.com/g', '')
    if (formattedDomain !== '') {
      formattedDomain = 'hylo.com/g/' + formattedDomain
    }
    return formattedDomain
  }

  removeUrlFromDomain (groupDomain) {
    return groupDomain.replace('hylo.com/g/', '')
  }

  render () {
    const { networkName } = this.props

    return <div styleName='flex-wrapper'>
      <ModalSidebar
        imageUrl={groovingAxolotl}
        onClick={this.props.goHome}
        theme={sidebarTheme}
        header='Everything looking good?'
        body='You can always come back and change your details at any time'
      />
      <div styleName='panel'>
        <div>
          <span styleName='step-count'>STEP 3/3</span>
        </div>
        <div styleName='center'>
          <div styleName='logo center' style={bgImageStyle(hyloNameWhiteBackground)} />
        </div>
        <div styleName='center-review'>
          <ReviewTextInput
            label={'Group Name'}
            value={this.state.edits.groupName || ''}
            readOnly={this.state.readOnly.groupName}
            editHandler={() => this.editHandler('groupName')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'groupName')}
            inputRef={this.inputRefs['groupName']}
          />
          <ReviewTextInput
            label={'URL'}
            value={formatDomainWithUrl(this.state.edits.groupDomain) || ''}
            readOnly={this.state.readOnly.groupDomain}
            editHandler={() => this.editHandler('groupDomain')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'groupDomain')}
            inputRef={this.inputRefs['groupDomain']}
          />
          {networkName && <ReviewTextInput
            label={'Network'}
            value={networkName}
            readOnly
          />}
          {/* <ReviewTextInput */}
          {/* label={'Privacy'} */}
          {/* value={this.state.edits.groupPrivacy} */}
          {/* onEnter={this.onEnter} */}
          {/* editHandler={() => this.props.goToPrivacyStep()} */}
          {/* /> */}

        </div>
        { this.state.error && <span styleName='review-arrow-up' /> }
        { this.state.error && <span styleName='review-error'>{this.state.error}</span>}
      </div>
      <ModalFooter
        submit={this.errorCheckAndSubmit}
        previous={this.previous}
        showPrevious={false}
        continueText={'Finish Up'}
      />
    </div>
  }
}

export function ReviewTextInput ({ label, value, editHandler, onChange, readOnly = true, inputRef }) {
  return <div styleName='review-input-text-row'>
    <div styleName='review-input-text-label'>
      <span>{label}</span>
    </div>
    <div styleName='review-input-text'>
      <TextInput
        type='text'
        name='group-name'
        value={value}
        theme={inputTheme}
        readOnly={readOnly}
        noClearButton
        onChange={onChange}
        inputRef={inputRef}
      />
    </div>
    {editHandler && <div styleName='review-input-edit'>
      <span styleName='edit-button' onClick={editHandler}>Edit</span>
    </div>}
  </div>
}

const inputTheme = {
  inputStyle: 'modal-input partial',
  wrapperStyle: 'center'
}

const sidebarTheme = {
  sidebarHeader: 'sidebar-header-full-page',
  sidebarText: 'gray-text sidebar-text-full-page'
}

const privacyOptions = [
  { id: '0', label: 'public' },
  { id: '1', label: 'private' },
  { id: '2', label: 'unlisted' }
]
