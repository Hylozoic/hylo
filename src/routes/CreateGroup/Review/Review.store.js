import orm from 'store/models'
import { createSelector as ormCreateSelector } from 'redux-orm'
import { AnalyticsEvents } from 'hylo-utils/constants'

export const MODULE_NAME = `Review`
export const CREATE_GROUP = `${MODULE_NAME}/CREATE_GROUP`

export function createGroup (name, slug, parentIds) {
  const data = {
    name,
    slug
  }

  if (parentIds) {
    data.parentIds = parentIds
  }

  return {
    type: CREATE_GROUP,
    graphql: {
      query: `mutation ($data: GroupInput) {
        createGroup(data: $data) {
          id
          hasModeratorRole
          group {
            id
            name
            slug
            parentGroups {
              items {
                id
              }
            }
          }
        }
      }
      `,
      variables: {
        data
      }
    },
    meta: {
      extractModel: 'Membership',
      ...data,
      analytics: AnalyticsEvents.GROUP_CREATED
    }
  }
}

export const getParents = ormCreateSelector(
  orm,
  (state, { parentIds }) => parentIds,
  (session, parentIds) => {
    return parentIds && parentIds.length > 0 ? session.Group.filter(g => parentIds.includes(g.id)).toModelArray() : []
  }
)
