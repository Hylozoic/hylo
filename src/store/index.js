import { createStore } from 'redux'
import createMiddleware from './middleware'
import { createBrowserHistory } from 'history'
import createRootReducer, { createCombinedReducers } from './reducers'

// `window` check here is for SSR case (see `appMiddleware`)
// where a static memory history is passed
export const history = typeof window !== 'undefined' && createBrowserHistory()

export function getEmptyState (providedHistory = history) {
  const combinedReducers = createCombinedReducers(providedHistory)

  return combinedReducers({}, { type: '' })
}

export default function configureStore (providedHistory = history, req) {
  return createStore(
    createRootReducer(providedHistory),
    getEmptyState(providedHistory),
    createMiddleware(providedHistory, req)
  )
}
