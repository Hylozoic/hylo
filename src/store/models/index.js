import { ORM } from 'redux-orm'
import './Model.extension'
import Activity from './Activity'
import Attachment from './Attachment'
import Comment from './Comment'
import EventInvitation from './EventInvitation'
import Group, { GroupConnection, GroupModerator, GroupJoinQuestion } from './Group'
import GroupTopic from './GroupTopic'
import Invitation from './Invitation'
import JoinRequest, { JoinRequestQuestionAnswer, Question } from './JoinRequest'
import LinkPreview from './LinkPreview'
import Location from './Location'
import Me, { MySkillsToLearn } from './Me'
import Membership from './Membership'
import Message from './Message'
import MessageThread from './MessageThread'
import Notification from './Notification'
import Person, { PersonSkillsToLearn } from './Person'
import PersonConnection from './PersonConnection'
import Post, { PostFollower, PostCommenter, ProjectMember } from './Post'
import PostMembership from './PostMembership'
import SearchResult from './SearchResult'
import Skill from './Skill'
import Topic from './Topic'
import Vote from './Vote'

export const orm = new ORM({ stateSelector: state => state.orm })

orm.register(
  Activity,
  Attachment,
  Comment,
  EventInvitation,
  Group,
  GroupConnection,
  GroupModerator,
  GroupJoinQuestion,
  GroupTopic,
  Invitation,
  LinkPreview,
  Location,
  JoinRequest,
  JoinRequestQuestionAnswer,
  Me,
  Membership,
  Message,
  MessageThread,
  MySkillsToLearn,
  Notification,
  Person,
  PersonConnection,
  PersonSkillsToLearn,
  Post,
  PostCommenter,
  PostFollower,
  PostMembership,
  ProjectMember,
  Question,
  SearchResult,
  Skill,
  Topic,
  Vote
)

export default orm
