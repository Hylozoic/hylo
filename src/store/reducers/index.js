import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'
import { history } from 'router'

import orm from './ormReducer'
import returnToPath from 'store/reducers/returnToPath'
import graphqlResponseError from 'store/reducers/graphqlResponseError'
import pending from './pending'
import locationHistory from './locationHistory'
import resetStore from './resetStore'
import mixpanel from './mixpanel'
import intercom from './intercom'
import queryResults from './queryResults'
import { handleSetState, composeReducers } from './util'

// Local store
// generator-marker-local-store-import
import AllTopics from 'routes/AllTopics/AllTopics.store'
import AttachmentManager from 'components/AttachmentManager/AttachmentManager.store'
import CreateGroup from 'components/CreateGroup/CreateGroup.store'
import CreateTopic from 'components/CreateTopic/CreateTopic.store'
import Events from 'routes/Events/Events.store'
import FeedList from 'components/FeedList/FeedList.store'
import FullPageModal from 'routes/FullPageModal/FullPageModal.store'
import HyloEditor from 'components/HyloEditor/HyloEditor.store'
import MapExplorer from 'routes/MapExplorer/MapExplorer.store'
import Messages from 'routes/Messages/Messages.store'
import Members from 'routes/Members/Members.store'
import MembershipRequests from 'routes/GroupSettings/MembershipRequestsTab/MembershipRequestsTab.store'
import MemberSelector from 'components/MemberSelector/MemberSelector.store'
import ModeratorsSettings from 'routes/GroupSettings/ModeratorsSettingsTab/ModeratorsSettingsTab.store'
import PeopleTyping from 'components/PeopleTyping/PeopleTyping.store'
import PostEditor from 'components/PostEditor/PostEditor.store'
import AuthLayoutRouter from 'routes/AuthLayoutRouter/AuthLayoutRouter.store'
import RelatedGroups from 'routes/GroupSettings/RelatedGroupsTab/RelatedGroupsTab.store'
import SavedSearches from 'routes/UserSettings/UserSettings.store'
import Search from 'routes/Search/Search.store'
import SkillsSection from 'components/SkillsSection/SkillsSection.store'
import SkillsToLearnSection from 'components/SkillsToLearnSection/SkillsToLearnSection.store'
import TopicsSettings from 'routes/GroupSettings/TopicsSettingsTab/TopicsSettingsTab.store'
import UserGroupsTab from 'routes/UserSettings/UserGroupsTab/UserGroupsTab.store'

export const combinedReducers = combineReducers({
  // Global store
  orm,
  router: connectRouter(history),
  returnToPath,
  graphqlResponseError,
  pending,
  queryResults,
  locationHistory,
  mixpanel,
  intercom,

  // Local store (Component)
  // generator-marker-local-store-reducer
  AllTopics,
  AttachmentManager,
  CreateGroup,
  CreateTopic,
  Events,
  FeedList,
  FullPageModal,
  HyloEditor,
  MembershipRequests,
  MapExplorer,
  Members,
  MemberSelector,
  Messages,
  ModeratorsSettings,
  PeopleTyping,
  AuthLayoutRouter,
  PostEditor,
  RelatedGroups,
  SavedSearches,
  Search,
  SkillsSection,
  SkillsToLearnSection,
  TopicsSettings,
  UserGroupsTab
})

export default composeReducers(
  combinedReducers,
  resetStore,
  handleSetState
)
