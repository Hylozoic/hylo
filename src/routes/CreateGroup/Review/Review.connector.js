import { get } from 'lodash/fp'
import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import getMe from 'store/selectors/getMe'
import updateUserSettings from 'store/actions/updateUserSettings'
import { addGroupName, addGroupDomain, fetchGroupExists } from '../CreateGroup.store'
import { createGroup, getParents } from './Review.store'

export function mapStateToProps (state, props) {
  const groupParentIds = get('parentIds', state.CreateGroup)
  const parentGroups = getParents(state, { parentIds: groupParentIds })

  return {
    currentUser: getMe(state),
    groupDomain: get('domain', state.CreateGroup),
    groupName: get('name', state.CreateGroup),
    groupPrivacy: get('privacy', state.CreateGroup),
    groupParentIds,
    groupDomainExists: get('domainExists', state.CreateGroup),
    groupParentNames: parentGroups.map(g => get('name', g))
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    goToGroup: (groupDomain) => dispatch(push(`/g/${groupDomain}`)),
    updateUserSettings: (changes) => dispatch(updateUserSettings(changes)),
    clearNameFromCreateGroup: () => dispatch(addGroupName(null)),
    clearDomainFromCreateGroup: () => dispatch(addGroupDomain(null)),
    createGroup: (name, slug, parentIds) => dispatch(createGroup(name, slug, parentIds)),
    goToPrivacyStep: () => dispatch(push('/create-group/privacy')),
    goHome: () => dispatch(push('/')),
    fetchGroupExists: (slug) => dispatch(fetchGroupExists(slug))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
