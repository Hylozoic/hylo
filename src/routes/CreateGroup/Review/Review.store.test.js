import { createGroup } from './Review.store'

describe('createGroup', () => {
  describe('without networkId', () => {
    it('omits networkId from variables', () => {
      const action = createGroup('thename', 'theslug')
      expect(action).toMatchSnapshot()
      expect(action.graphql.variables.data.networkId).toEqual(undefined)
    })
  })

  // TODO: fix up with parents
  // describe('with parentIds', () => {
  //   it('includes networkId in variables', () => {
  //     const networkId = 123
  //     const action = createGroup('thename', 'theslug', networkId)
  //     expect(action).toMatchSnapshot()
  //     expect(action.graphql.variables.data.networkId).toEqual(networkId)
  //   })
  // })
})
