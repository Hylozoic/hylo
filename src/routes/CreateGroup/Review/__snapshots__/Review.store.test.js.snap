// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGroup with networkId includes networkId in variables 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($data: GroupInput) {
        createGroup(data: $data) {
          id
          hasModeratorRole
          group {
            id
            name
            slug
          }
        }
      }
      ",
    "variables": Object {
      "data": Object {
        "name": "thename",
        "networkId": 123,
        "slug": "theslug",
      },
    },
  },
  "meta": Object {
    "analytics": "Group Created",
    "extractModel": "Membership",
    "name": "thename",
    "networkId": 123,
    "slug": "theslug",
  },
  "type": "Review/CREATE_GROUP",
}
`;

exports[`createGroup without networkId omits networkId from variables 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($data: GroupInput) {
        createGroup(data: $data) {
          id
          hasModeratorRole
          group {
            id
            name
            slug
          }
        }
      }
      ",
    "variables": Object {
      "data": Object {
        "name": "thename",
        "slug": "theslug",
      },
    },
  },
  "meta": Object {
    "analytics": "Group Created",
    "extractModel": "Membership",
    "name": "thename",
    "slug": "theslug",
  },
  "type": "Review/CREATE_GROUP",
}
`;
